### ISSUE TITLE
Python Interface - Can't step and run in a breakpoint handler

### ISSUE THREAD

**Issue by JeremyWildsmith** [x64dbg/x64dbg#1915]
- Debugger version: FEB 14 2018
- Operating system: Windows 10
- Brief description of the issue: threading model makes it very difficult / impossible to work with breakpoint handlers
- Elaborate reproduction steps for the bug/issue being reported.

I am using the Python x64dbgpy plugin and have a very simple breakpoint handler that looks like the following:
````
def bphit_callback():
    global traceAreaArray
    global installAddr

    if installAddr == Register.EIP:
        installBreakpoints()

    for ta in traceAreaArray:
        if ta.jumpAddress == Register.EIP:
            
            x64dbg.DbgCmdExecDirect("StepInto")      
            branchAddressStr = hex(getattr(Register, "EIP"))
            ta.hitCount = ta.hitCount + 1
            if branchAddressStr not in ta.branches:
                ta.branches.append(branchAddressStr)

    x64dbg.DbgCmdExecDirect("run")
````
Essentially I am trying to map out the flow control of an obfuscated application by tracing the jump instructions (which as assembled as jmp [some register], or ret)

The above code will, once a breakpoint is hit (which is on a flow control instruction) step into and then record EIP to detect where the subroutine jumped to. However, due to threading model of x64dbg it seems like it is not possible to step and run within a breakpoint handler

Now I can't think of a way to do this at all with the interface I have; is there any way I can do this? I've tried having the bp handler onto another thread which was invoked by the bp callback routine but that doesn't seem to work either.

**Comment by JeremyWildsmith** [comment]
Okay,

I've found a way around this by not using callbacks and just looping Debug.Run and Debug.StepInto etc but x64dbg keeps crashing about a minute into the script running. Hmm... Not really sure why.

**Comment by mrexodia** [comment]
Hm, strange... another solution is to use the DbgGetBranchDestination
inside the callback.

Could you send an executable and script to reproduce the crash?

**Comment by torusrxxx** [comment]
Conditional breakpoint doesn't allow any stepping in the plugin callback or
breakpoint command.

**Comment by JeremyWildsmith** [comment]
I cannot legally redistribute the executable but I can provide details on where to get it. The script is here:
https://github.com/JeremyWildsmith/x64dbg_scripts/blob/master/traceAreas.py

**Comment by mrexodia** [comment]
Yeah please provide details on where to get it. The `DbgBranchDestination` never failed for me personally, you can find the implementation here maybe you can spot an issue https://github.com/x64dbg/x64dbg/blob/development/src/dbg/_exports.cpp#L742

**Comment by JeremyWildsmith** [comment]
Yeah these crashes are making it really hard to run analysis that run longer than 5 minutes. I am going to try to reproduce it on a redistributable binary first.
